src/gui/window.rs

```diff
@@ -47,82 +52,254 @@ impl Fps {
+        // and return its height in a variable
+        let panel_height = if temp_settings.panel_visible {
+            let panel_height = 40.0;
+            let panel_frame = Frame::default()
+                .inner_margin(egui::style::Margin::same(10.0))
+                .fill(egui::Color32::from_rgb(42, 42, 42));
+            egui::TopBottomPanel::top("Top panel")
+                .height_range(panel_height..=panel_height)
+                .frame(panel_frame)
+                .show(&ctx, |ui| {
+                    ui.horizontal(|ui| {
+                        if ui.button("Open MIDI").clicked() {
+                            let midi_path = FileDialog::new()
+                                .add_filter("midi", &["mid"])
+                                .set_directory("/")
+                                .pick_file();
 
-                if ui.small_button("Back 5 sec").clicked() {
-                    if time >= five_sec {
-                        self.midi_file.timer_mut().seek(time - five_sec);
+                            if let Some(midi_path) = midi_path {
+                                let mut midi_file =
+                                    MIDIFileUnion::InRam(InRamMIDIFile::load_from_file(
+                                        &midi_path.into_os_string().into_string().unwrap(),
+                                        SimpleTemporaryPlayer::new(&perm_settings.sfz_path),
+                                        perm_settings.random_colors,
+                                    ));
+                                midi_file.timer_mut().play();
+                                self.midi_file = Some(midi_file);
+                            }
+                        }
+                        if self.midi_file.is_some() && ui.button("Close MIDI").clicked() {
+                            self.midi_file = None;
+                        }
+                        if ui.button("Play").clicked() {
+                            self.midi_file.as_mut().unwrap().timer_mut().play();
+                        }
+                        if ui.button("Pause").clicked() {
+                            self.midi_file.as_mut().unwrap().timer_mut().pause();
+                        }
+                        if ui.button("Settings").clicked() {
+                            match temp_settings.settings_visible {
+                                true => temp_settings.settings_visible = false,
+                                false => temp_settings.settings_visible = true,
+                            }
+                        }
+                        ui.horizontal(|ui| {
+                            ui.label("Note speed: ");
+                            ui.add(
+                                egui::Slider::new(&mut perm_settings.note_speed, 2.0..=0.001)
+                                    .show_value(false),
+                            );
+                        })
+                    });
+
+                    if self.midi_file.is_some() {
```

Comment:
I will always be picky about this because even though this wouldn't crash it's still good to avoid .unwrap() as much as possible.

You could do something like:
```rs
if let Some(midi_file) = self.midi_file.as_mut() {
```
or
```rs
if let Some(midi_file) = &mut self.midi_file {
```
then you don't need the 3 unwraps inside.

Reply:
Somehow I keep missing those unwraps
Will fix